editor/editor_data.cpp

```diff
@@ -1141,6 +1141,27 @@ Ref<Texture2D> EditorData::get_script_icon(const Ref<Script> &p_script) {
 		base_scr = base_scr->get_base_script();
 	}
 
+	// No custom icon was found in the inheritance chain, so check the base
+	// class of the script instead.
+	String base_type;
+	p_script->get_language()->get_global_class_name(p_script->get_path(), &base_type);
+
+	// Check if the base type is an extension-defined type.
+	Ref<Texture2D> ext_icon = extension_class_get_icon(base_type);
+	if (ext_icon.is_valid()) {
+		_script_icon_cache[p_script] = ext_icon;
+		return ext_icon;
+	}
+
+	// Look for the base type in the editor theme.
+	// This is only relevant for built-in classes.
+	const Control *gui_base = EditorNode::get_singleton()->get_gui_base();
+	if (gui_base && gui_base->has_theme_icon(base_type, SNAME("EditorIcons"))) {
+		Ref<Texture2D> theme_icon = gui_base->get_theme_icon(base_type, SNAME("EditorIcons"));
+		_script_icon_cache[p_script] = theme_icon;
+		return theme_icon;
+	}
+
```

Comment:
Problem with caching this is that adding icons to an existing class would require an editor restart. Though I guess changing an icon already requires an editor restart. We need a way to invalidate this cache. I guess this could be fine for now.

Would you mind adding a TODO comment around here somewhere pointing that limitation out and mentioning that we need to invalidate the cache when relevant?

Reply:
Currently the cache is cleared when closing a scene tab.

Reply:
Uh-huh. Well that explains why things are so slow. We try to rely on this cache everywhere in the editor, but the moment any scene is closed, we just flush all of it. That's not great. I now realize that I have touched this code, but missed the implication of it.